---
title: "Untitled"
author: "RÃ©mi Legrand"
date: "2023-04-26"
output: html_document
---



# Try with some better filters on the blast

```{r}
hit_table <- read.csv('../../data/new_assign/HitTable.csv', header = F, 
                      col.names = c("seq_num", "seq_id", "percent_identity", "alignment_length", "mismatches", "gap_opens", "q.start", "q.end", "s.start", "s.end", "evalue", "bit_score"))

hit_table
```

# Lowest e-val

Here, I wanted to get the smallest and I there is a few, the set of the lowest

```{r}
# function to get the smallest e-val depending on the sequence
min_eval <- function(seq_id){
  min(hit_table[hit_table$seq_num == seq_id,"evalue"])
}

# Gather it in a df with the smallest with the smallest e-val for the sequences (not optimized at all because for loop but easier to read)

identified <- data.frame()

for (i in unique(hit_table$seq_num)) {
  identified <- rbind(identified, hit_table[hit_table$seq_num == i & hit_table$evalue == min_eval(i),])
  # print(i)
}

# Get all the unique id of the best matching sequences (around 4-5% of all the matching sequences)

# unique(identified$seq_id)

# If we say that all the found sequences are relevant, then we can just omit the previous step and obtain

# unique(hit_table$seq_id)

# Percent of the kept sequences in the first case:

length(unique(hit_table$seq_id))/length(unique(identified$seq_id))*100
```

# Recover the sequences to make a FASTA file


```{r}
# library(biomaRt)
# 
# # download the genome of Homo sapiens from refseq
# # and store the corresponding genome file in '_ncbi_downloads/genomes'
# HS.genome.refseq <- getGenome( db       = "refseq", 
#                                organism = "Homo sapiens")
# # import downloaded genome as Biostrings object
# Human_Genome <- read_genome(file = HS.genome.refseq)
# # look at the Biostrings object
# Human_Genome
```

```{r}
# # install.packages("readxl")
# library(readxl)
# 
# ref <- read_xlsx("../../data/nifH_dada2_phylum_v2.0.2.xlsx")
# 
# length(unique(ref$tax2))
# length(ref$acc)
# 
# library(tidyverse)
# phylogeny[1,]
# phylogeny <- as_tibble(ref$tax2)
# tax <- separate(phylogeny, value, sep = ";", into = c("cluster", "domain", "phylum", "class", "order", "family", "genus"))
# tax
# length(unique(tax$genus))
# 
# length(which(is.na(tax$class)))
```

```{r}
new_db <- read.csv('../../data/new_assign/sequence.ipg.csv', header = T, sep = '\t')
# view(new_db)
```


```{r}
ncbi_accession <- new_db[new_db$Source == "RefSeq",]
ncbi_accession

length(unique(ncbi_accession$Id))

fileConn<-file("../../data/new_assign/refseq_accession.txt")
writeLines(ncbi_accession$Nucleotide.Accession, fileConn)
close(fileConn)

uni <- (unique(as.character(ncbi_accession$Id)))

fileConn<-file("../../data/new_assign/refseq_id")
writeLines(uni, fileConn)
close(fileConn)
```

# Try to access with R entrez

```{r}
#  install.packages("rentrez")
```

```{r}
# library(rentrez)
# # entrez_set_email("remi.legrand38@gmail.com")
# accession_numbers <- ncbi_accession$Nucleotide.Accession[1:10]
# class(accession_numbers)
# 
# # Install the rentrez package if not already installed
# install.packages("rentrez")

# Load the rentrez package
library(rentrez)

# Define the list of accession numbers
accession_numbers <- c("NZ_FQZT01000024", "NZ_CP045300", "NZ_FPAU01000001")

# Fetch the data for each accession number
for (accession in accession_numbers) {
  # Search for the protein record
  search_term <- paste0(accession, "[Accession]")
  search_result <- entrez_search(db = "protein", term = search_term)
  
  # Get the protein record ID
  record_id <- search_result$ids[1]
  
  # Fetch the protein record
  protein_record <- entrez_fetch(db = "protein", id = record_id, rettype = "gbwithparts", retmode = "text")
  
  # Extract the DNA sequence and taxonomy information
  dna_sequence <- getSequence(protein_record)
  taxonomy <- getTaxonomy(protein_record)
  
  # Print the results
  cat("Accession:", accession, "\n")
  cat("DNA Sequence:\n", dna_sequence, "\n")
  cat("Taxonomy:", taxonomy, "\n\n")
}

# Function to extract the DNA sequence from the protein record
getSequence <- function(record) {
  pattern <- "ORIGIN(.+)//"
  match <- regmatches(record, regexpr(pattern, record, perl = TRUE))
  
  if (length(match) > 0) {
    sequence <- gsub("\\s+|\\d+|//", "", match)
    return(sequence)
  } else {
    return("DNA sequence not found.")
  }
}

# Function to extract the taxonomy information from the protein record
getTaxonomy <- function(record) {
  pattern <- "ORGANISM(.+)?REFERENCE"
  match <- regmatches(record, regexpr(pattern, record, perl = TRUE))
  
  if (length(match) > 0) {
    taxonomy <- gsub("\\s+", " ", match)
    return(taxonomy)
  } else {
    return("Taxonomy information not found.")
  }
}


```

